uncertainties/core.py

```diff
@@ -2766,6 +2779,22 @@ def std_dev(self, std_dev):
 
         self._std_dev = float(std_dev)
 
+    def __hash__(self):
+        """
+        Calculates the hash for any Variable object.
+        This simply calls the __hash__ method of `AffineScalarFunc` in most cases.
+        During initialization, the `_linear_part` and `_nominal_value` attributes does not exist.
+        Because of that, the id returned in case this method is called during initialization.
+        
+        Returns:
+            int: The hash of this object
+        """
+
+        if hasattr(self, '_linear_part') and hasattr(self, '_nominal_value'):
+            return super().__hash__()
+        else:
+            return id(self)
```

Comment:
The test failure is coming from this part. `__hash__` has to give the same value every time. To avoid the error, one option is to change `__hash__` here to:

    return hash((self._nominal_value, (id(self),), (1.,)))

and update  `__init__` to set `self._nominal_value` before the `{self: 1.}` dict is created. `__setstate__` also needs to be updated to make sure `_nominal_value` is set before it creates the self-referential dict (but see my other comment about wanting to avoid tuples).

It is tempting to try to remove the self-reference but I have seen another case in a symbolic algebra library where a set up similar to this one was the best option.

Reply:
You are totally right, the hash must be the same every time. Not sure, how I could miss that.